### PR TITLE
Add adversarial loss

### PR DESCRIPTION
**Pull request by grenlayk**

### CODE REVIEW

File: src/config/stylegan_adversarial.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            disk.download(weights_folder, weights_folder)
+
+        model_G = StyleBased_Generator(dim_latent=512)
+        model_G.load_state_dict(torch.load(f'{weights_folder}/model'))
+        model_G.to(device)
+
+        style_embedder = models.resnet18()
+        style_embedder.fc = torch.nn.Identity()
+        style_embedder = style_embedder.to(device)
+        style_embedder.load_state_dict(torch.load(f'{weights_folder}/style_embedder'))
```
Comment (SphericalPotatoInVacuum) [inline src/config/stylegan_adversarial.py:47]: We moved content embedder to a separate file, and as I understand it style embedder also kinda is a modified resnet, maybe we should make it a separate model too?

File: src/training/stylegan_adversarial.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        valid = torch.tensor(1.).expand_as(pred_D_real).to(self.device)
+        loss_D_real = self.adv_loss(pred_D_real, valid) 
+
+        return (loss_D_real + loss_D_fake) * 0.5
+
+    def model_G_adv_loss(self, res: torch.Tensor):
+        pred_D_fake = self.model_D(res)
+        valid = torch.tensor(1.).expand_as(pred_D_fake).to(self.device)
+
+        return self.adv_loss(pred_D_fake, valid)
```
Comment (SphericalPotatoInVacuum) [inline src/training/stylegan_adversarial.py:92]: These could be a separate loss object maybe

**Review by SphericalPotatoInVacuum — APPROVED** [review]

File: src/training/stylegan_adversarial.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            # update models
+            self.set_requires_grad(self.model_D, True)
+
+            loss_G.backward()
+            loss_D.backward()
+
+            self.optimizer_D.step() 
+            self.optimizer_G.step()
+            self.scheduler_D.step()
+            self.scheduler_G.step()
```
Comment (SphericalPotatoInVacuum) [inline src/training/stylegan_adversarial.py:164]: These should be done after each epoch

**Review by SphericalPotatoInVacuum — CHANGES_REQUESTED** [review]

File: src/config/stylegan_adversarial.py (hunk 1 above)
Comment (nikoryagin) [reply src/config/stylegan_adversarial.py:47]: But do we really want to move 2 lines to other file? I think it’s okay to leave it like it is

File: src/config/stylegan_adversarial.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+        content_embedder = ContentResnet(BasicBlock, [2, 2, 2, 2]).to(device)
```
[the remark this answers is not in the document]
Comment (nikoryagin) [reply src/config/stylegan_adversarial.py:49]: https://github.com/grenlayk/text-deep-fake/pull/16/files#diff-42be64e0fe92723953407641a429f9c90f80fd0a828c3da9f6cf185e011495d9R48 same here

File: src/training/stylegan_adversarial.py (hunk 2 above)
Comment (nikoryagin) [reply src/training/stylegan_adversarial.py:92]: Agreed

File: src/config/stylegan_adversarial.py (hunk 1 above)
Comment (grenlayk) [reply src/config/stylegan_adversarial.py:47]: We can move it to `embedder.py` file.

File: src/training/stylegan_adversarial.py (hunk 2 above)
Comment (grenlayk) [reply src/training/stylegan_adversarial.py:92]: not sure because we need params of the trainer in these funcs

**Review by SphericalPotatoInVacuum — APPROVED** [review]

File: src/training/stylegan_adversarial.py (hunk 2 above)
Comment (nikoryagin) [reply src/training/stylegan_adversarial.py:92]: True, I think it’s okay both ways then, it’s up to you

Comment (SphericalPotatoInVacuum) [reply src/training/stylegan_adversarial.py:92]: We could pass them along the batches, or when constructing the loss. But yeah, it's not a big deal.

File: src/config/stylegan_adversarial.py (hunk 1 above)
Comment (nikoryagin) [reply src/config/stylegan_adversarial.py:47]: I would leave it here, but if you want to move it, then we should definitely move it to  embedder.py

**Review by nikoryagin — APPROVED** [review]

**Review by nikoryagin — APPROVED** [review]
